08-optional-types.Rmd

```diff
@@ -0,0 +1,344 @@
+
+### Making an entire task optional
+Let's say we want to make an entire task optional. Perhaps it provides additional QC information, or isn't relevent to certain data types.
+
+
+Although WDL has a concept of `if`, it does not have any concept of `else`, so to make mutually exclusive tasks, you need to do something like this:
+```
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+}
+```
+
+However, we need to be aware that WDL is not "aware" that these two tasks are mutually exclusive. You and I both know that if task_x has run, then task_y must not have run, but your WDL executor does not know this. So, as a consequence, you cannot use mutually exclusive if-blocks to set a variable to a particular value.
+
+```
+# this workflow will fail miniwdl check or womtool
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        Int some_variable = 10  # this is problematic, even though only one of these if blocks will execute!
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        Int some_variable = 5  # this is problematic, even though only one of these if blocks will execute!
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+}
+```
+Another consequence is that both of your mutually exclusive tasks' outputs will be considered optional types, even though you can be certain that set of outputs absolutely do exist and one set of outputs absolutely do not exist.
+
+```
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+
+    Boolean did_x_run = defined(task_x.something)
+    Boolean did_y_run = defined(task_y.something)
+
+    outputs {
+        # these both MUST be optional
+        File? task_x_output = task_x.something
+        File? task_y_output = task_y.something
+    }
+}
+```
+
+## Checking if a variable is defined
+You can check if a variable is defined at runtime using the WDL built-in `defined()`, which returns a Boolean value. However, `defined()` can give unexpected output, as we will go over later. For the time being, let's use a simple example.
+
+```
+workflow do_x_or_y {
```

Comment:
This seems to be an almost identical code chunk example as above? Maybe just use one and put the writing together?

Reply:
Oops, copy error! Thanks for spotting this, they were supposed to be more different.